applications/IgaApplication/custom_modelers/refinement_modeler.cpp

```diff
@@ -0,0 +1,197 @@
+                    p_nurbs_surface->Spans(spans_local_space, 1);
+
+                    std::vector<double> knots_to_insert_v;
+                    for (IndexType i = 0; i < spans_local_space.size() - 1; ++i) {
+                        const double delta_v = (spans_local_space[i] - spans_local_space[i + 1]) / (nb_per_span_v + 1);
+                        for (IndexType j = 0; j < spans_local_space.size() - 1; ++j) {
+                            knots_to_insert_v.push_back(spans_local_space[j] + delta_v);
+                        }
+                    }
+
+                    PointerVector<NodeType> PointsRefined;
+                    Vector KnotsURefined;
+                    Vector WeightsRefined;
+
+                    NurbsSurfaceRefinementUtilities::KnotRefinementU(*(p_nurbs_surface.get()), knots_to_insert_v,
+                        PointsRefined, KnotsURefined, WeightsRefined);
+
+                    // Recreate nodes in model part to ensure correct assignment of dofs
+                    IndexType node_id = (r_model_part.NodesEnd() - 1)->Id() + 1;
+                    for (IndexType i = 0; i < PointsRefined.size(); ++i) {
+                        if (PointsRefined(i)->Id() == 0) {
+                            PointsRefined(i) = r_model_part.CreateNewNode(node_id, PointsRefined[i][0], PointsRefined[i][1], PointsRefined[i][2]);
+                            node_id++;
+                        }
+                    }
+
+                    p_nurbs_surface->SetInternals(PointsRefined,
+                        p_nurbs_surface->PolynomialDegreeU(), p_nurbs_surface->PolynomialDegreeV(),
+                        KnotsURefined, p_nurbs_surface->KnotsV(),
+                        WeightsRefined);
+                }
+            }
+        }
+    }
+
+    void RefinementModeler::GetGeometryList(
+        GeometriesArrayType& rGeometryList,
+        ModelPart& rModelPart,
+        const Parameters rParameters) const
+    {
+        if (rParameters.Has("brep_id")) {
+            rGeometryList.push_back(rModelPart.pGetGeometry(rParameters["brep_id"].GetInt()));
+        }
+        if (rParameters.Has("brep_ids")) {
+            for (SizeType i = 0; i < rParameters["brep_ids"].size(); ++i) {
+                rGeometryList.push_back(rModelPart.pGetGeometry(rParameters["brep_ids"][i].GetInt()));
+            }
+        }
+        if (rParameters.Has("brep_name")) {
+            rGeometryList.push_back(rModelPart.pGetGeometry(rParameters["brep_name"].GetString()));
+        }
+        if (rParameters.Has("brep_names")) {
+            for (SizeType i = 0; i < rParameters["brep_names"].size(); ++i) {
+                rGeometryList.push_back(rModelPart.pGetGeometry(rParameters["brep_names"][i].GetString()));
+            }
+        }
+
+        KRATOS_ERROR_IF(rGeometryList.size() == 0)
+            << "Empty geometry list. Either \"brep_id\", \"brep_ids\", \"brep_name\" or \"brep_names\" are the possible options." << std::endl;
```

Comment:
here I would then add volume_id. We then just have to make sure that the nurbs_geometry_modeler is called before the refinement modeler.